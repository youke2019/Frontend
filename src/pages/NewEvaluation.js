import React from "react";
import {
    Text,
    View,
    TextInput,
    StyleSheet,
    ScrollView,
    TouchableOpacity,
} from "react-native";
import Rating from '../components/Rating'
import axios from 'axios'
import Toast from 'react-native-easy-toast'
import StackNavBar from "../components/StackNavBar";
import {Image} from "react-native-elements";

class NewEvaluation extends React.Component {
    state = {
        post: {
            credit_point: null,
        },
        descriptions: [],
        count: 0
    }


    setData = (type, text) => {
        let data = {}
        data[type] = text
        this.setState({
            post: Object.assign({},this.state.post,data)
        })
    }

    setDescription = (text,index,type) => {
        let descriptions = this.state.descriptions
        let description = this.state.descriptions[index]

        if (type == 'title')
            description.title = text
        if (type == 'content')
            description.content = text

        descriptions[index] = description

        this.setState({
            descriptions: descriptions
        })
    }

    postEvaluation = () => {
        if (
            this.state.post.credit_point == null
            || this.state.post.授课教师 == null
            || this.state.post.老师评价 == null
            || this.state.post.考核形式及难度 == null
            || this.state.post.课程个人体验 == null
        ){
            this.refs.toast.show('请打分并填写必填部分')
        } else {
            let data={
                user_id: this.props.navigation.state.params.user_id,
                course_id: this.props.navigation.state.params.course_info.course_id
            }

            for (let description of this.state.descriptions)
                if (description.title != null && description.content != null)
                    data[description.title] = description.content

            data = Object.assign({}, data, this.state.post)

            axios.post(baseUrl+'/courses/evaluates/add', data).then(()=>{
                this.props.navigation.navigate("Evaluations")
            }).catch(err=>{
                console.log(err)
            })

            console.log(data)
        }
    }

    newDescription = () => {
        console.log(this.state.count)
        let descriptions = this.state.descriptions
        descriptions.push({
            title: null,
            content: null,
            count: this.state.count++
        })
        this.setState({
            descriptions: descriptions
        })
    }

    deleteDescription = (index) => {
        let descriptions = this.state.descriptions
        descriptions.splice(index,1)
        this.setState({
            descriptions: descriptions
        })
    }

    render() {
        const {
            descriptions
        } = this.state

        return (
            <View>
                <ScrollView>
                    <Toast ref="toast"/>
                    <StackNavBar
                        navigation={this.props.navigation}
                        title='发布评测'
                        buttonText='发布'
                        onPress={this.postEvaluation}
                    />
                    <View  style={styles.container}>
                        <View style={styles.star_container}>
                            <Text style={[styles.star_text,styles.title]}>请对课程进行评分</Text>
                            <Rating
                                selected={true}
                                onUpdate={(rating) => this.setData('credit_point',rating)}
                            />
                        </View>
                        <View style={[styles.card_container,styles.border]}>
                            <Text style={styles.title}>授课教师</Text>
                            <View style={styles.input_container}>
                                <TextInput
                                    placeholder={"输入授课教师(必填)"}
                                    multiline={true}
                                    style={styles.input}
                                    onChangeText={(text) => this.setData('授课教师',text)}
                                />
                            </View>
                        </View>
                        <View style={[styles.card_container,styles.border]}>
                            <Text style={styles.title}>老师评价</Text>
                            <View style={styles.input_container}>
                                <TextInput
                                    placeholder={"输入老师评价(必填)"}
                                    multiline={true}
                                    style={styles.input}
                                    onChangeText={(text) => this.setData('老师评价',text)}
                                />
                            </View>
                        </View>
                        <View style={[styles.card_container,styles.border]}>
                            <Text style={styles.title}>考核形式及难度</Text>
                            <View style={styles.input_container}>
                                <TextInput
                                    placeholder={"输入考核形式及难度(必填)"}
                                    multiline={true}
                                    style={styles.input}
                                    onChangeText={(text) => this.setData('考核形式及难度',text)}
                                />
                            </View>
                        </View>
                        <View style={[styles.card_container,styles.border,styles.bottom]}>
                            <Text style={styles.title}>课程个人体验</Text>
                            <View style={styles.input_container}>
                                <TextInput
                                    placeholder={"输入课程个人体验(必填)"}
                                    multiline={true}
                                    style={styles.input}
                                    onChangeText={(text) => this.setData('课程个人体验',text)}
                                />
                            </View>
                        </View>
                        {
                            descriptions.map((item,index) => {
                                return(
                                    <View
                                        style={styles.description}
                                        key={item.count}
                                    >
                                        <TouchableOpacity
                                            style={styles.delete}
                                            onPress={() => this.deleteDescription(index)}
                                        >
                                            <Image
                                                source={{uri:'close'}}
                                                style={styles.delete_icon}
                                            />
                                        </TouchableOpacity>
                                        <View
                                            style={[styles.card_container,styles.border]}
                                        >
                                            <TextInput
                                                style={styles.title}
                                                placeholder={"输入描述标题"}
                                                onChangeText={(text) => this.setDescription(text,index,'title')}
                                            />
                                            <View style={styles.input_container}>
                                                <TextInput
                                                    placeholder={"输入描述内容"}
                                                    multiline={true}
                                                    style={styles.input}
                                                    onChangeText={(text) => this.setDescription(text,index,'content')}
                                                />
                                            </View>
                                        </View>
                                    </View>
                                )
                            })
                        }
                    </View>
                </ScrollView>
                <View
                    style={styles.post_container}
                >
                    <TouchableOpacity
                        style={styles.post_button}
                        onPress={this.newDescription}
                    >
                        <Text style={styles.post_text}>添加新的描述</Text>
                    </TouchableOpacity>
                </View>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container:{
        backgroundColor: '#FFFFFA',
        paddingBottom: 120,
    },
    border:{
        margin: 25,
        borderRadius: 10,
        elevation: 3,
        backgroundColor: '#FFFFFF',
    },
    star_container:{
        padding: 36,
    },
    star_text:{
        padding: 10,
    },
    card_container:{
        flex:1,
        paddingHorizontal: 20,
        paddingVertical: 20,
    },
    input_container:{

    },
    description:{
        flexDirection: 'row',
        alignItems: 'center'
    },
    delete:{
        paddingLeft: 18,
    },
    delete_icon:{
        width: 25,
        height: 25,
    },
    title:{
        color: 'black',
        fontSize: 17,
    },
    input:{
        padding: 16,
        fontSize: 14,
    },
    post_container:{
        justifyContent: 'center',
        alignItems: 'center',
        top: -60,
        flex:1,
    },
    post_button:{
        backgroundColor: '#FDAF26',
        paddingHorizontal: 30,
        paddingVertical: 10,
        borderRadius: 50,
        height: 42,
        width: 180,
        justifyContent: 'center',
        alignItems: 'center',
    },
    post_text:{
        fontSize: 16,
    },
})

export default NewEvaluation