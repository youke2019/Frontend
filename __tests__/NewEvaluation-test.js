import React from 'react'
import renderer from 'react-test-renderer'
import {shallow, mount, render, configure} from 'enzyme'
import NewEvaluation from '../src/pages/NewEvaluation'
import Adapter from "enzyme-adapter-react-16/build"

configure({adapter: new Adapter()})

const navigation = {
    state:{
        params:{
            user_id: "test",
            course_info:{
                course_id: "test",
            }
        }
    },
    navigate: jest.fn(()=>{}),
    goBack: jest.fn(()=>{})
}

console.error = jest.fn()
jest.mock('axios')
global.baseUrl = 'baseUrl'

describe('functions', () => {

    let wrapper = mount(
        <NewEvaluation
            navigation={navigation}
        />)

    it('setData()', () => {
        wrapper.instance().setData('testType','testText')
        expect(wrapper.instance().state.post).toHaveProperty('testType','testText')
    })

    it('postEvaluation() abnormal', (done) => {
        wrapper.setState({
            post: {
                credit_point: -1,
                课程简述: 'testOutline',
            }
        })
        wrapper.instance().postEvaluation()
        const promise = new Promise(
            function (resolve, reject)
            {resolve('success')})
        promise
            .then(data=>{
                expect(navigation.navigate).not.toHaveBeenCalled()
                done()
            })
            .catch(err=>{console.log(err)})
    })

    it('postEvaluation() normal', (done) => {
        wrapper.setState({
            post: {
                credit_point: null,
                课程简述: null,
            }
        })

        wrapper.instance().postEvaluation()

        wrapper.setState({
            post: {
                credit_point: 8,
                课程简述: 'testOutline',
            }
        })
        wrapper.instance().postEvaluation()
        const promise = new Promise(
            function (resolve, reject)
            {resolve('success')})
        promise
            .then(data=>{
                expect(navigation.navigate).toHaveBeenCalled()
                done()
            })
            .catch(err=>{console.log(err)})
    })

    it('anonymous functions', () => {
        wrapper.find({placeholder:'输入课程简述'}).at(0).props().onChangeText('test')
        expect(wrapper.instance().state.post).toHaveProperty('课程简述','test')

        wrapper.find({placeholder:'输入考核形式'}).at(0).props().onChangeText('test')
        expect(wrapper.instance().state.post).toHaveProperty('考核形式','test')

        wrapper.find({placeholder:'输入上课自由程度'}).at(0).props().onChangeText('test')
        expect(wrapper.instance().state.post).toHaveProperty('上课自由程度','test')

        wrapper.find({placeholder:'输入课程个人体验'}).at(0).props().onChangeText('test')
        expect(wrapper.instance().state.post).toHaveProperty('课程个人体验','test')

        wrapper.find('Rating').props().onUpdate(6)
        expect(wrapper.instance().state.post).toHaveProperty('credit_point',6)

    })
})

test('renders QuestionCard page correctly', () => {
    const tree = renderer.create(
        <NewEvaluation
        />).toJSON()
    expect(tree).toMatchSnapshot();
})