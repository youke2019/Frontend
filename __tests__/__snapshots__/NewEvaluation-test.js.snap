// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders QuestionCard page correctly 1`] = `
<View>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 40,
            "justifyContent": "center",
            "marginLeft": 10,
            "marginTop": 24,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "height": "100%",
              "justifyContent": "center",
              "opacity": 1,
              "paddingHorizontal": 6,
            }
          }
        >
          <Image
            resizeMode="contain"
            source={
              Object {
                "uri": "navigate_return",
              }
            }
            style={
              Object {
                "height": 20,
                "width": 20,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 10,
              "flexDirection": "row",
              "justifyContent": "flex-start",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 16,
                "paddingLeft": 8,
              }
            }
            theme={
              Object {
                "colors": Object {
                  "disabled": "hsl(208, 8%, 90%)",
                  "divider": "#bcbbc1",
                  "error": "#ff190c",
                  "grey0": "#393e42",
                  "grey1": "#43484d",
                  "grey2": "#5e6977",
                  "grey3": "#86939e",
                  "grey4": "#bdc6cf",
                  "grey5": "#e1e8ee",
                  "greyOutline": "#bbb",
                  "platform": Object {
                    "android": Object {
                      "error": "#f44336",
                      "primary": "#2196f3",
                      "secondary": "#9C27B0",
                      "success": "#4caf50",
                      "warning": "#ffeb3b",
                    },
                    "ios": Object {
                      "error": "#ff3b30",
                      "primary": "#007aff",
                      "secondary": "#5856d6",
                      "success": "#4cd964",
                      "warning": "#ffcc00",
                    },
                  },
                  "primary": "#2089dc",
                  "searchBg": "#303337",
                  "secondary": "#8F0CE8",
                  "success": "#52c41a",
                  "warning": "#faad14",
                },
              }
            }
          >
            发布评测
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FDAF26",
              "borderRadius": 20,
              "height": 32,
              "justifyContent": "center",
              "marginRight": 10,
              "opacity": 1,
              "width": 80,
            }
          }
        >
          <Text
            style={Object {}}
            theme={
              Object {
                "colors": Object {
                  "disabled": "hsl(208, 8%, 90%)",
                  "divider": "#bcbbc1",
                  "error": "#ff190c",
                  "grey0": "#393e42",
                  "grey1": "#43484d",
                  "grey2": "#5e6977",
                  "grey3": "#86939e",
                  "grey4": "#bdc6cf",
                  "grey5": "#e1e8ee",
                  "greyOutline": "#bbb",
                  "platform": Object {
                    "android": Object {
                      "error": "#f44336",
                      "primary": "#2196f3",
                      "secondary": "#9C27B0",
                      "success": "#4caf50",
                      "warning": "#ffeb3b",
                    },
                    "ios": Object {
                      "error": "#ff3b30",
                      "primary": "#007aff",
                      "secondary": "#5856d6",
                      "success": "#4cd964",
                      "warning": "#ffcc00",
                    },
                  },
                  "primary": "#2089dc",
                  "searchBg": "#303337",
                  "secondary": "#8F0CE8",
                  "success": "#52c41a",
                  "warning": "#faad14",
                },
              }
            }
          >
            发布
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFA",
            "paddingBottom": 120,
          }
        }
      >
        <View
          style={
            Object {
              "padding": 36,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "padding": 10,
                },
                Object {
                  "color": "black",
                  "fontSize": 17,
                },
              ]
            }
          >
            请对课程进行评分
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              accessible={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "paddingHorizontal": 5,
                }
              }
            >
              <View
                accessibilityIgnoresInvertColors={true}
                style={
                  Object {
                    "height": 25,
                    "width": 25,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "uri": "star1",
                    }
                  }
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                      Object {
                        "height": 25,
                        "width": 25,
                      },
                      Object {
                        "resizeMode": "cover",
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              accessible={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "paddingHorizontal": 5,
                }
              }
            >
              <View
                accessibilityIgnoresInvertColors={true}
                style={
                  Object {
                    "height": 25,
                    "width": 25,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "uri": "star2",
                    }
                  }
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                      Object {
                        "height": 25,
                        "width": 25,
                      },
                      Object {
                        "resizeMode": "cover",
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              accessible={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "paddingHorizontal": 5,
                }
              }
            >
              <View
                accessibilityIgnoresInvertColors={true}
                style={
                  Object {
                    "height": 25,
                    "width": 25,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "uri": "star3",
                    }
                  }
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                      Object {
                        "height": 25,
                        "width": 25,
                      },
                      Object {
                        "resizeMode": "cover",
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              accessible={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "paddingHorizontal": 5,
                }
              }
            >
              <View
                accessibilityIgnoresInvertColors={true}
                style={
                  Object {
                    "height": 25,
                    "width": 25,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "uri": "star4",
                    }
                  }
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                      Object {
                        "height": 25,
                        "width": 25,
                      },
                      Object {
                        "resizeMode": "cover",
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              accessible={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "paddingHorizontal": 5,
                }
              }
            >
              <View
                accessibilityIgnoresInvertColors={true}
                style={
                  Object {
                    "height": 25,
                    "width": 25,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "uri": "star5",
                    }
                  }
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                      Object {
                        "height": 25,
                        "width": 25,
                      },
                      Object {
                        "resizeMode": "cover",
                      },
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "paddingHorizontal": 20,
                "paddingVertical": 20,
              },
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 10,
                "elevation": 3,
                "margin": 25,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 17,
              }
            }
          >
            授课教师
          </Text>
          <View
            style={Object {}}
          >
            <TextInput
              allowFontScaling={true}
              multiline={true}
              onChangeText={[Function]}
              placeholder="输入授课教师(必填)"
              rejectResponderTermination={true}
              style={
                Object {
                  "fontSize": 14,
                  "padding": 16,
                }
              }
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "paddingHorizontal": 20,
                "paddingVertical": 20,
              },
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 10,
                "elevation": 3,
                "margin": 25,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 17,
              }
            }
          >
            老师评价
          </Text>
          <View
            style={Object {}}
          >
            <TextInput
              allowFontScaling={true}
              multiline={true}
              onChangeText={[Function]}
              placeholder="输入老师评价(必填)"
              rejectResponderTermination={true}
              style={
                Object {
                  "fontSize": 14,
                  "padding": 16,
                }
              }
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "paddingHorizontal": 20,
                "paddingVertical": 20,
              },
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 10,
                "elevation": 3,
                "margin": 25,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 17,
              }
            }
          >
            考核形式及难度
          </Text>
          <View
            style={Object {}}
          >
            <TextInput
              allowFontScaling={true}
              multiline={true}
              onChangeText={[Function]}
              placeholder="输入考核形式及难度(必填)"
              rejectResponderTermination={true}
              style={
                Object {
                  "fontSize": 14,
                  "padding": 16,
                }
              }
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "paddingHorizontal": 20,
                "paddingVertical": 20,
              },
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 10,
                "elevation": 3,
                "margin": 25,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 17,
              }
            }
          >
            课程个人体验
          </Text>
          <View
            style={Object {}}
          >
            <TextInput
              allowFontScaling={true}
              multiline={true}
              onChangeText={[Function]}
              placeholder="输入课程个人体验(必填)"
              rejectResponderTermination={true}
              style={
                Object {
                  "fontSize": 14,
                  "padding": 16,
                }
              }
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "top": -60,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FDAF26",
          "borderRadius": 50,
          "height": 42,
          "justifyContent": "center",
          "opacity": 1,
          "paddingHorizontal": 30,
          "paddingVertical": 10,
          "width": 180,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 16,
          }
        }
      >
        添加新的描述
      </Text>
    </View>
  </View>
</View>
`;
